OutOfMemoryException | Unable to create new native threads error

Using the latest extension randomly gives java.lang.OutOfMemoryError: unable to create new native thread. 
This happens when the ES indexing job extension based jobs is executed multiple times in DC. 
This does not happen with the version 2.6 where the ElasticSearchDatastores are not used but instead a local ElasticSearchClientFactory is used. 

Stacktrace :-
66) Error injecting constructor, java.lang.OutOfMemoryError: unable to create new native thread
  at org.elasticsearch.threadpool.ThreadPool.<init>(Unknown Source)
  while locating org.elasticsearch.threadpool.ThreadPool
  for parameter 3 at org.elasticsearch.client.transport.TransportClientNodesService.<init>(Unknown Source)
  while locating org.elasticsearch.client.transport.TransportClientNodesService
Caused by: java.lang.OutOfMemoryError: unable to create new native thread
        at java.lang.Thread.start0(Native Method)
        at java.lang.Thread.start(Thread.java:714)
        at org.elasticsearch.threadpool.ThreadPool.<init>(ThreadPool.java:159)
        at sun.reflect.GeneratedConstructorAccessor411.newInstance(Unknown Source)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
        at org.elasticsearch.common.inject.DefaultConstructionProxyFactory$1.newInstance(DefaultConstructionProxyFactory.java:54)
        at org.elasticsearch.common.inject.ConstructorInjector.construct(ConstructorInjector.java:86)
        at org.elasticsearch.common.inject.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:98)
        at org.elasticsearch.common.inject.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:45)
        at org.elasticsearch.common.inject.InjectorImpl.callInContext(InjectorImpl.java:837)
        at org.elasticsearch.common.inject.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:42)
        at org.elasticsearch.common.inject.Scopes$1$1.get(Scopes.java:57)
        at org.elasticsearch.common.inject.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:45)
        at org.elasticsearch.common.inject.SingleParameterInjector.inject(SingleParameterInjector.java:42)
        at org.elasticsearch.common.inject.SingleParameterInjector.getAll(SingleParameterInjector.java:66)
        at org.elasticsearch.common.inject.ConstructorInjector.construct(ConstructorInjector.java:85)
        at org.elasticsearch.common.inject.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:98)
        at org.elasticsearch.common.inject.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:45)
        at org.elasticsearch.common.inject.InjectorImpl.callInContext(InjectorImpl.java:837)
        at org.elasticsearch.common.inject.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:42)
        at org.elasticsearch.common.inject.Scopes$1$1.get(Scopes.java:57)
        at org.elasticsearch.common.inject.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:45)
        at org.elasticsearch.common.inject.InjectorBuilder$1.call(InjectorBuilder.java:200)
        at org.elasticsearch.common.inject.InjectorBuilder$1.call(InjectorBuilder.java:193)
        at org.elasticsearch.common.inject.InjectorImpl.callInContext(InjectorImpl.java:830)
        at org.elasticsearch.common.inject.InjectorBuilder.loadEagerSingletons(InjectorBuilder.java:193)
        at org.elasticsearch.common.inject.InjectorBuilder.injectDynamically(InjectorBuilder.java:175)
        at org.elasticsearch.common.inject.InjectorBuilder.build(InjectorBuilder.java:110)
        at org.elasticsearch.common.inject.Guice.createInjector(Guice.java:93)
        at org.elasticsearch.common.inject.Guice.createInjector(Guice.java:70)
        at org.elasticsearch.common.inject.ModulesBuilder.createInjector(ModulesBuilder.java:59)
        at org.elasticsearch.client.transport.TransportClient.<init>(TransportClient.java:187)
        at org.elasticsearch.client.transport.TransportClient.<init>(TransportClient.java:123)
        at org.datacleaner.connection.ElasticSearchDatastore.createDatastoreConnection(ElasticSearchDatastore.java:127)
        at org.datacleaner.connection.UsageAwareDatastore.getDatastoreConnection(UsageAwareDatastore.java:117)
        at org.datacleaner.connection.UsageAwareDatastore.openConnection(UsageAwareDatastore.java:128)
        at org.datacleaner.connection.ElasticSearchDatastore.openConnection(ElasticSearchDatastore.java:151)
        at org.datacleaner.extension.elasticsearch.ElasticSearchIndexAnalyzer.init(ElasticSearchIndexAnalyzer.java:89)
        at sun.reflect.GeneratedMethodAccessor174.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.datacleaner.descriptors.AbstractMethodDescriptor.invoke(AbstractMethodDescriptor.java:79)
        at org.datacleaner.descriptors.InitializeMethodDescriptorImpl.initialize(InitializeMethodDescriptorImpl.java:44)
        at org.datacleaner.lifecycle.LifeCycleHelper.initialize(LifeCycleHelper.java:180)
        at org.datacleaner.job.tasks.InitializeTask.executeInternal(InitializeTask.java:83)
        at org.datacleaner.job.tasks.InitializeTask.execute(InitializeTask.java:55)
        at org.datacleaner.job.concurrent.TaskRunnable.run(TaskRunnable.java:61)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

--- a/src/main/java/org/datacleaner/extension/elasticsearch/ElasticSearchDocumentIdLookupTransformer.java
+++ b/src/main/java/org/datacleaner/extension/elasticsearch/ElasticSearchDocumentIdLookupTransformer.java
@@ -22,9 +22,12 @@ package org.datacleaner.extension.elasticsearch;
 import javax.inject.Named;
 
 import org.apache.metamodel.elasticsearch.ElasticSearchDataContext;
+import org.apache.metamodel.util.FileHelper;
 import org.datacleaner.api.Categorized;
+import org.datacleaner.api.Close;
 import org.datacleaner.api.Configured;
 import org.datacleaner.api.Description;
+import org.datacleaner.api.Initialize;
 import org.datacleaner.api.InputColumn;
 import org.datacleaner.api.InputRow;
 import org.datacleaner.api.OutputColumns;
@@ -65,6 +68,21 @@ public class ElasticSearchDocumentIdLookupTransformer implements ElasticSearchTr
     @Description("Fields to return")
     String[] fields;
 
+    private UpdateableDatastoreConnection _connection;
+
+    @Initialize
+    public void init() {
+        _connection = elasticsearchDatastore.openConnection();
+    }
+
+    @Close
+    public void close() {
+        if (_connection != null) {
+            FileHelper.safeClose(_connection);
+            _connection = null;
+        }
+    }
+
     @Override
     public OutputColumns getOutputColumns() {
         return new OutputColumns(String.class, fields);
@@ -73,8 +91,8 @@ public class ElasticSearchDocumentIdLookupTransformer implements ElasticSearchTr
     @Override
     public String[] transform(InputRow row) {
 
-        try (UpdateableDatastoreConnection connection = elasticsearchDatastore.openConnection()) {
-            final ElasticSearchDataContext dataContext = (ElasticSearchDataContext) connection.getDataContext();
+        try {
+            final ElasticSearchDataContext dataContext = (ElasticSearchDataContext) _connection.getDataContext();
 
             final Client client = dataContext.getElasticSearchClient();
             final String[] result = new String[fields.length];

--- a/src/main/java/org/datacleaner/extension/elasticsearch/ElasticSearchFullSearchTransformer.java
+++ b/src/main/java/org/datacleaner/extension/elasticsearch/ElasticSearchFullSearchTransformer.java
@@ -24,9 +24,12 @@ import java.util.Map;
 import javax.inject.Named;
 
 import org.apache.metamodel.elasticsearch.ElasticSearchDataContext;
+import org.apache.metamodel.util.FileHelper;
 import org.datacleaner.api.Categorized;
+import org.datacleaner.api.Close;
 import org.datacleaner.api.Configured;
 import org.datacleaner.api.Description;
+import org.datacleaner.api.Initialize;
 import org.datacleaner.api.InputColumn;
 import org.datacleaner.api.InputRow;
 import org.datacleaner.api.OutputColumns;
@@ -69,6 +72,21 @@ public class ElasticSearchFullSearchTransformer implements ElasticSearchTransfor
 
     @Configured(order = 4, required = false)
     String searchFieldName;
+    
+    private UpdateableDatastoreConnection _connection;
+    
+    @Initialize
+    public void init() {
+        _connection = elasticsearchDatastore.openConnection();
+    }
+    
+    @Close
+    public void close() {
+        if (_connection != null) {
+            FileHelper.safeClose(_connection);
+            _connection = null;
+        }
+    }
 
     @Override
     public OutputColumns getOutputColumns() {
@@ -85,8 +103,8 @@ public class ElasticSearchFullSearchTransformer implements ElasticSearchTransfor
         if (StringUtils.isNullOrEmpty(input)) {
             return result;
         }
-        try (UpdateableDatastoreConnection connection = elasticsearchDatastore.openConnection()) {
-            final ElasticSearchDataContext dataContext = (ElasticSearchDataContext) connection.getDataContext();
+        try {
+            final ElasticSearchDataContext dataContext = (ElasticSearchDataContext) _connection.getDataContext();
             final Client client = dataContext.getElasticSearchClient();
             MatchQueryBuilder query;
             if (StringUtils.isNullOrEmpty(searchFieldName)) {

--- a/src/main/java/org/datacleaner/extension/elasticsearch/ElasticSearchIndexAnalyzer.java
+++ b/src/main/java/org/datacleaner/extension/elasticsearch/ElasticSearchIndexAnalyzer.java
@@ -19,14 +19,15 @@
  */
 package org.datacleaner.extension.elasticsearch;
 
-import java.io.IOException;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import javax.inject.Named;
 
 import org.apache.metamodel.elasticsearch.ElasticSearchDataContext;
+import org.apache.metamodel.util.FileHelper;
 import org.datacleaner.api.Analyzer;
 import org.datacleaner.api.Categorized;
+import org.datacleaner.api.Close;
 import org.datacleaner.api.Configured;
 import org.datacleaner.api.Description;
 import org.datacleaner.api.Initialize;
@@ -82,17 +83,19 @@ public class ElasticSearchIndexAnalyzer implements Analyzer<WriteDataResult> {
 
     private AtomicInteger _counter;
     private WriteBuffer _writeBuffer;
+    private UpdateableDatastoreConnection _connection;
 
     @Initialize
-    public void init() throws IOException {
+    public void init() throws Exception {
+        _connection = elasticsearchDatastore.openConnection();
 
-        try (UpdateableDatastoreConnection connection = elasticsearchDatastore.openConnection()) {
-            final ElasticSearchDataContext dataContext = (ElasticSearchDataContext) connection.getDataContext();
+        try {
+            final ElasticSearchDataContext dataContext = (ElasticSearchDataContext) _connection.getDataContext();
 
             final Client client = dataContext.getElasticSearchClient();
             _counter = new AtomicInteger(0);
-            _writeBuffer = new WriteBuffer(bulkIndexSize, new ElasticSearchIndexFlushAction(elasticsearchDatastore,
-                    fields, documentType));
+            _writeBuffer = new WriteBuffer(bulkIndexSize, new ElasticSearchIndexFlushAction(dataContext, fields,
+                    documentType));
 
             final String indexName = elasticsearchDatastore.getIndexName();
 
@@ -105,7 +108,16 @@ public class ElasticSearchIndexAnalyzer implements Analyzer<WriteDataResult> {
                     .actionGet();
         } catch (Exception e) {
             logger.error("Exception while running the ElasticSearchIndexAnalyzer", e);
+            FileHelper.safeClose(_connection);
             throw e;
+        }
+    }
+
+    @Close
+    public void close() {
+        if (_connection != null) {
+            FileHelper.safeClose(_connection);
+            _connection = null;
         }
     }
 

--- a/src/main/java/org/datacleaner/extension/elasticsearch/ElasticSearchIndexFlushAction.java
+++ b/src/main/java/org/datacleaner/extension/elasticsearch/ElasticSearchIndexFlushAction.java
@@ -25,8 +25,6 @@ import java.util.Map;
 
 import org.apache.metamodel.elasticsearch.ElasticSearchDataContext;
 import org.apache.metamodel.util.Action;
-import org.datacleaner.connection.ElasticSearchDatastore;
-import org.datacleaner.connection.UpdateableDatastoreConnection;
 import org.elasticsearch.action.bulk.BulkRequestBuilder;
 import org.elasticsearch.action.bulk.BulkResponse;
 import org.elasticsearch.action.index.IndexRequest;
@@ -43,21 +41,19 @@ public class ElasticSearchIndexFlushAction implements Action<Iterable<Object[]>>
     private static final Logger logger = LoggerFactory.getLogger(ElasticSearchIndexFlushAction.class);
     private final String[] _fields;
     private final String _documentType;
-    private final ElasticSearchDatastore _elasticSearchDatastore;
+    private final ElasticSearchDataContext _elasticSearchDataContext;
 
-    public ElasticSearchIndexFlushAction(ElasticSearchDatastore elasticSearchDatastore, String[] fields,
+    public ElasticSearchIndexFlushAction(ElasticSearchDataContext elasticSearchDataContext, String[] fields,
             String documentType) {
-        _elasticSearchDatastore = elasticSearchDatastore;
+        _elasticSearchDataContext = elasticSearchDataContext;
         _fields = fields;
         _documentType = documentType;
     }
 
     @Override
     public void run(Iterable<Object[]> rows) throws Exception {
-
-        try (UpdateableDatastoreConnection connection = _elasticSearchDatastore.openConnection()) {
-            final ElasticSearchDataContext dataContext = (ElasticSearchDataContext) connection.getDataContext();
-            final Client client = dataContext.getElasticSearchClient();
+        try {
+            final Client client = _elasticSearchDataContext.getElasticSearchClient();
             final BulkRequestBuilder bulkRequestBuilder = new BulkRequestBuilder(client);
 
             for (Object[] row : rows) {
@@ -77,7 +73,7 @@ public class ElasticSearchIndexFlushAction implements Action<Iterable<Object[]>>
                     }
                 }
                 logger.debug("Indexing record ({}): {}", id, map);
-                final IndexRequest indexRequest = new IndexRequest(_elasticSearchDatastore.getIndexName(),
+                final IndexRequest indexRequest = new IndexRequest(_elasticSearchDataContext.getIndexName(),
                         _documentType, id);
                 indexRequest.source(map);
                 indexRequest.operationThreaded(false);

--- a/src/test/java/org/datacleaner/extension/elasticsearch/ElasticSearchDocumentIdLookupTransformerTest.java
+++ b/src/test/java/org/datacleaner/extension/elasticsearch/ElasticSearchDocumentIdLookupTransformerTest.java
@@ -62,6 +62,7 @@ public class ElasticSearchDocumentIdLookupTransformerTest extends TestCase {
         OutputColumns out = transformer.getOutputColumns();
         assertEquals("OutputColumns[city, country]", out.toString());
 
+        transformer.init();
         try {
 
             _server.truncateIndex();
@@ -85,8 +86,9 @@ public class ElasticSearchDocumentIdLookupTransformerTest extends TestCase {
             assertEquals("[Delhi, India]", Arrays.toString(output));
 
         } catch (Exception e) {
-            e.printStackTrace();
             throw e;
+        } finally {
+            transformer.close();
         }
 
     }

--- a/src/test/java/org/datacleaner/extension/elasticsearch/ElasticSearchDocumentIdLookupTransformerTest.java
+++ b/src/test/java/org/datacleaner/extension/elasticsearch/ElasticSearchDocumentIdLookupTransformerTest.java
@@ -64,7 +64,6 @@ public class ElasticSearchDocumentIdLookupTransformerTest extends TestCase {
 
         transformer.init();
         try {
-
             _server.truncateIndex();
             assertEquals(0, _server.getDocumentCount());
 
@@ -84,9 +83,6 @@ public class ElasticSearchDocumentIdLookupTransformerTest extends TestCase {
 
             output = transformer.transform(new MockInputRow().put(col1, "del"));
             assertEquals("[Delhi, India]", Arrays.toString(output));
-
-        } catch (Exception e) {
-            throw e;
         } finally {
             transformer.close();
         }

--- a/src/test/java/org/datacleaner/extension/elasticsearch/ElasticSearchFullSearchTransformerTest.java
+++ b/src/test/java/org/datacleaner/extension/elasticsearch/ElasticSearchFullSearchTransformerTest.java
@@ -61,10 +61,12 @@ public class ElasticSearchFullSearchTransformerTest extends TestCase {
         transformer.searchInput = col1;
         transformer.documentType = ElasticSearchTestServer.DOCUMENT_TYPE;
         transformer.elasticsearchDatastore = _elasticSearchDatastore;
+        
 
         OutputColumns out = transformer.getOutputColumns();
         assertEquals("OutputColumns[Document ID, Document]", out.toString());
 
+        transformer.init();
         try {
 
             _server.truncateIndex();
@@ -91,8 +93,9 @@ public class ElasticSearchFullSearchTransformerTest extends TestCase {
             assertEquals("null", String.valueOf(output[1]));
 
         } catch (Exception e) {
-            e.printStackTrace();
             throw e;
+        } finally {
+            transformer.close();
         }
     }
 }

--- a/src/test/java/org/datacleaner/extension/elasticsearch/ElasticSearchFullSearchTransformerTest.java
+++ b/src/test/java/org/datacleaner/extension/elasticsearch/ElasticSearchFullSearchTransformerTest.java
@@ -68,7 +68,6 @@ public class ElasticSearchFullSearchTransformerTest extends TestCase {
 
         transformer.init();
         try {
-
             _server.truncateIndex();
             assertEquals(0, _server.getDocumentCount());
 
@@ -91,9 +90,6 @@ public class ElasticSearchFullSearchTransformerTest extends TestCase {
             output = transformer.transform(new MockInputRow().put(col1, "n/a"));
             assertEquals("null", String.valueOf(output[0]));
             assertEquals("null", String.valueOf(output[1]));
-
-        } catch (Exception e) {
-            throw e;
         } finally {
             transformer.close();
         }

